pallets/providers/src/utils.rs

```diff
@@ -2267,4 +2384,186 @@ where
+            if block_to_clean > NextStartingBlockToCleanUp::<T>::get() {
+                NextStartingBlockToCleanUp::<T>::put(block_to_clean);
+                meter.consume(db_weight.writes(1));
+            }
+
+            meter
+        }
+
+        fn process_block_expired_items(meter: &mut WeightMeter) {
+            let db_weight = T::DbWeight::get();
+            let minimum_required_weight_processing_expired_items = db_weight.reads_writes(2, 1);
+
+            // Check if there is enough remaining weight to process expired move bucket requests
+            if !meter.can_consume(minimum_required_weight_processing_expired_items) {
+                return;
+            }
+
+            // Provider top up expirations expire based on the relay chain block number, not the local block number.
+            // TODO: use next starting block to clean up instead of current block number here
+            let relay_chain_block_number = RelayBlockGetter::<T>::current_block_number();
+
+            // Remove expired move bucket requests if any existed and process them.
+            let mut provider_top_up_expirations =
+                ProviderTopUpExpirations::<T>::take(&relay_chain_block_number);
+            meter.consume(minimum_required_weight_processing_expired_items);
+
+            // TODO: After benchmarking, we should check before this loop that there is enough remaining weight to
+            // TODO: process all the expired move bucket requests. If not, we should return early.
+            while let Some(provider_id) = provider_top_up_expirations.pop() {
+                Self::process_expired_provider_top_up_period(provider_id, meter);
+            }
+
+            // If there are remaining items which were not processed, put them back in storage
+            if !provider_top_up_expirations.is_empty() {
+                ProviderTopUpExpirations::<T>::insert(
+                    &relay_chain_block_number,
+                    provider_top_up_expirations,
+                );
+                meter.consume(db_weight.writes(1));
+            }
+        }
+
+        fn process_expired_provider_top_up_period(
+            provider_id: ProviderIdFor<T>,
+            meter: &mut WeightMeter,
+        ) {
+            let db_weight = T::DbWeight::get();
+            let potential_weight = db_weight.reads_writes(0, 2);
+
+            if !meter.can_consume(potential_weight) {
+                return;
+            }
+
+            // Clear awaiting top up storage
+            let maybe_awaiting_top_up = AwaitingTopUpFromProviders::<T>::take(&provider_id);
+
+            // Mark the provider as insolvent if it was awaiting a top up
+            // If the provider was not awaiting a top up, it means they already topped up either via an
+            // automatic top up or a manual top up.
+            let current_tick =
+                <T::PaymentStreams as shp_traits::PaymentStreamsInterface>::current_tick();
```

Comment:
I'm of the opinion of simplifying things here and just not allow Providers to charge after they've been marked as insolvent. Period. Not letting them charge up to the point they were solvent, just not letting them charge at all. Cause now you have to involve the Payment Streams pallet tick in the Providers pallet.

Reply:
so i modified it so that they can't charge while their insolvent, but also as we discussed, users will now only be charged for the time that the provider was not awaiting for top ups.

Reply:
Why storing this information if Providers are not going to be able to charge at all.